runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,482 @@
+  local extmark_id = vim.api.nvim_buf_set_extmark(bufnr, snippet_ns, range[1], range[2], {
+    right_gravity = false,
+    end_right_gravity = true,
+    end_line = range[3],
+    end_col = range[4],
+  })
+
+  local self = setmetatable(
+    { index = index, bufnr = bufnr, extmark_id = extmark_id },
+    { __index = Tabstop }
+  )
+
+  return self
+end
+
+--- Returns the tabstop's range.
+---
+--- @return Range4
+function Tabstop:get_range()
+  local mark =
+    vim.api.nvim_buf_get_extmark_by_id(self.bufnr, snippet_ns, self.extmark_id, { details = true })
+
+  --- @diagnostic disable-next-line: undefined-field
+  return { mark[1], mark[2], mark[3].end_row, mark[3].end_col }
+end
+
+--- Returns the text spanned by the tabstop.
+---
+--- @return string
+function Tabstop:get_text()
+  local range = self:get_range()
+  local lines = vim.api.nvim_buf_get_lines(self.bufnr, range[1], range[3] + 1, false)
+  lines[#lines] = lines[#lines]:sub(1, range[4])
+  lines[1] = lines[1]:sub(range[2] + 1)
+
+  return table.concat(lines, '\n')
+end
+
+--- Sets the tabstop's text.
+---
+--- @param text string
+function Tabstop:set_text(text)
+  local range = self:get_range()
+  vim.api.nvim_buf_set_text(
+    self.bufnr,
+    range[1],
+    range[2],
+    range[3],
+    range[4],
+    remove_newlines(text)
+  )
+end
+
+--- @class vim.snippet.Session
+--- @field bufnr integer
+--- @field tabstops table<integer, vim.snippet.Tabstop[]>
+--- @field current_tabstop vim.snippet.Tabstop
+local Session = {}
+
+--- Initializes a new snippet session in the current buffer.
+---
+--- @return vim.snippet.Session
+function Session:init()
+  self.bufnr = vim.api.nvim_get_current_buf()
+  self.tabstops = {}
+  self.current_tabstop = Tabstop.new(0, self.bufnr, { 0, 0, 0, 0 })
+
+  return self
+end
+
+--- Returns the number of tabstop indexes in the snippet session.
+---
+--- @return integer
+function Session:get_num_tabstops()
+  return #vim.tbl_keys(self.tabstops)
+end
+
+--- Adds a tabstop at the given index and spanning `range`.
+---
+--- @param index integer
+--- @param range Range4
+function Session:add_tabstop(index, range)
+  self.tabstops[index] = self.tabstops[index] or {}
+  table.insert(self.tabstops[index], Tabstop.new(index, self.bufnr, range))
+end
+
+--- @class vim.snippet.Snippet
+--- @field private session? vim.snippet.Session
+local M = { session = nil }
+
+--- Select the given tabstop range.
+---
+--- @private
+--- @param tabstop vim.snippet.Tabstop
+function M._select_stop(tabstop)
+  --- @param keys string
+  local function feedkeys(keys)
+    keys = vim.api.nvim_replace_termcodes(keys, true, false, true)
+    vim.api.nvim_feedkeys(keys, 'n', true)
+  end
+
+  --- @param row integer
+  --- @param col integer
+  local function move_cursor_to(row, col)
+    local line = vim.fn.getline(row) --[[ @as string ]]
+    col = math.max(vim.fn.strchars(line:sub(1, col)) - 1, 0)
+    feedkeys(string.format('%sG0%s', row, string.rep('<Right>', col)))
+  end
```

Comment:
Would `api.nvim_win_set_cursor` also work here? If not - maybe add a comment why it can't be used

Reply:
Good question. I initially tried that but for some reason that causes the cursor to end up at the end of the selection instead of the beginning. I used [`nvim-snippy`](https://github.com/dcampos/nvim-snippy/blob/5787f088c99c04309a8aa6363fd3ac9a5a26b780/lua/snippy/main.lua#L39-L53) as a reference to handle tabstop selection and this logic is taken from there.

Reply:
I've left the requested comment.